### PR TITLE
Fix Choose menu label when a menu has been deleted.

### PR DESCRIPTION
**Pull request by afercia**
Fixes part of https://github.com/WordPress/gutenberg/issues/59431

## What?

See https://github.com/WordPress/gutenberg/issues/59431#issuecomment-2475906971

When a menu has been deleted, the settings panel shows a warning Notice. In that state, the ellipsis button to choose a menu is unlabeled. The condition to set the dynamic label fails and doesn't take into account this scenario.

Screenshot 2024-11-14 at 10 59 43

## Why?

All controls should always be properly labeled.

## How?

It appears the condition to render the Norice:

https://github.com/WordPress/gutenberg/blob/a120f10e63e3c6eb5fa4970e8ab3690e74453fde/packages/block-library/src/navigation/edit/menu-inspector-controls.js#L96

is slightly different from the condition(s) used to determine the ellipsis button label:

https://github.com/WordPress/gutenberg/blob/a120f10e63e3c6eb5fa4970e8ab3690e74453fde/packages/block-library/src/navigation/edit/navigation-menu-selector.js#L105-L117

Specifically in this scenario, `currentMenuId && isNavigationMenuMissing` is true, while `menuUnavailable` is false.

I'm not familiar with this part to determine whether `menuUnavailable` is incorrect in the first place and whether it should be reviewed. For now, I'm just adding one more check, which is the same one used for the Notice, to render the label 'Choose or create a Navigation Menu'. Any feedback on the current `menuUnavailable` would be appreciated.

## Testing Instructions

- Make sure you have a navigation menu assigned to the Navigation block.
- Go to navigation panel in the Site Editor > Navigation > select the assigned menu, and delete it from the ellipsis menu.
- In the editor canvas, edit the navigation block (you'll likely have to switch the editor to template editing mode).
- Select the Navigation block.
- In the settings panel, observe a Notice is rendered.
- Hover or focus the ellipsis button above the Notice.
- Observe the ellipsis button shows a tooltip 'Choose or create a Navigation Menu'.
- Inspect the source and observe the button has an aria-label with the same text.

### CODE REVIEW

**Comment by afercia** [comment]
Attempting to fix the failing test. I have a question:
maybe I'm missing something, why the menu ID wasn't passed to `useNavigationMenu()`? Isn't it supposed to always receive the menu ID?
Cc @WordPress/gutenberg-core

File: packages/block-library/src/navigation/edit/navigation-menu-selector.js [hunk 1]
```diff
@@ -106,12 +107,18 @@ function NavigationMenuSelector( {
 	const noBlockMenus = ! hasNavigationMenus && hasResolvedNavigationMenus;
 	const menuUnavailable =
 		hasResolvedNavigationMenus && currentMenuId === null;
+	const navMenuHasBeenDeleted = currentMenuId && isNavigationMenuMissing;
```
Comment (ntsekouras) [inline packages/block-library/src/navigation/edit/navigation-menu-selector.js:110]: In trunk we have some similar checks and others with `hasResolvedNavigationMenus && isNavigationMenuMissing`. I'm wondering why that difference and mostly why isn't `isNavigationMenuMissing` is doing this check internally in the hook.

@afercia 's comment about `Specifically in this scenario, currentMenuId && isNavigationMenuMissing is true, while menuUnavailable is false.` and why we weren't passing the id here are good ones. Maybe @getdave or @draganescu have more context?

Comment (draganescu) [reply packages/block-library/src/navigation/edit/navigation-menu-selector.js:110]: > why isn't isNavigationMenuMissing is doing this check internally in the hook

I assume you mean check for hasResolvedNavigationMenus? One of the main problems with the code of the navigation block is that it is a blueprint of all the pitfalls we've ran through as the block evolved. I remember these checks Is* and has* have been part of a "refactoring" to herd some of the repeating code.

> currentMenuId && isNavigationMenuMissing

This check is very valid, I am unsure what the question is, it means: the block has a navigation id assigned but it's missing.

Comment (ntsekouras) [reply packages/block-library/src/navigation/edit/navigation-menu-selector.js:110]: The first question is how `currentMenuId && isNavigationMenuMissing` and `hasResolvedNavigationMenus && isNavigationMenuMissing` are different. Related to this is why `isNavigationMenuMissing` not checking this internally, so we wouldn't need the first part of the above checks.

The other two from @afercia are:
1. Why currentMenuId && isNavigationMenuMissing is true, while menuUnavailable is false (this is mentioned in the PR's description.
2. [why we weren't passing the id](https://github.com/WordPress/gutenberg/pull/67009#issuecomment-2476706438)

Comment (draganescu) [reply packages/block-library/src/navigation/edit/navigation-menu-selector.js:110]: I guess the easiest way to clear things up are via code, I find it unlikely somebody knows that by heart 🤷🏻 Particularly the `menuUnavailable` bit, which sounds like a bug?

As to why we're not passing the `id` I think the main reason is it wasn't meant to and probably evolved in a way in which it now seems that it should.

**Comment by afercia** [comment]
> why we weren't passing the id

@ntsekouras by reading [this test name](https://github.com/WordPress/gutenberg/blob/91b130b2db14028768e8e85fdf83ac8e5cb0b5b8/packages/block-library/src/navigation/test/use-navigation-menu.js#L173) `Should return information about all menus when the ref is missing` and [the PR](https://github.com/WordPress/gutenberg/pull/41139) that refactored the `useNavigationMenu` hook, specifically [this comment](https://github.com/WordPress/gutenberg/pull/41139#pullrequestreview-979781397). maybe I'm understanding what passing the ID does.

To my understanding, despite its name, the `useNavigationMenu` hook returns information about _all_ menus _and_ about the menu with the passed ID. The very confusing thing is that some returned values change depending on whether the ID is passed. If the ID is missing, `isNavigationMenuMissing` [is always true](https://github.com/WordPress/gutenberg/blob/91b130b2db14028768e8e85fdf83ac8e5cb0b5b8/packages/block-library/src/navigation/use-navigation-menu.js#L83-L89).

In this PR, we need to check whether the menu is missing or not so that, to my understanding, we need to pass the ID.

As an aside, I'd think the documentation of the hook is unclear and should be improved. It would be good to make clear that the ID is necessary to get reliable values about the current menu. 

On the other hand, I'm not sure returning values anyways even when the ID is missing makes the `useNavigationMenu()` API well understandable. For example, as mentioned above, when the ID is not passed, `isNavigationMenuMissing` is assumed to be true which is confusing. At the very least, I'd consider to document this behavior, in a separate issue and PR.

Any thoughts and a new review welcome 🙏🏻

**Comment by afercia** [comment]
Rebased.

**Comment by afercia** [comment]
I created https://github.com/WordPress/gutenberg/issues/67828 to consider some audit and better documentation of the `useNavigationMenu` hook. In the meantime, I'd think this PR is ready for a review Cc @WordPress/gutenberg-core when you have a chance.

**Review by ntsekouras — APPROVED** [review]
Nice sleuthing @afercia! That API could use some simplifications..

This looks good, thank you!

**Comment by afercia** [comment]
@ntsekouras thanks for your review!

**Pull request merged**